emcc.py

```diff
@@ -2725,6 +2725,8 @@ def parse_args(newargs):
   options = EmccOptions()
   settings_changes = []
   should_exit = False
+  eh_enabled = False
+  wasm_eh_enabled = False
```

Comment:
why add these two variables? shouldn't the info all be in the `Settings` values?

Reply:
`eh_enabled` is set when `-fexceptions` is specified, and `wasm_eh_enabled` is set when `-fwasm-exceptions` is specified. `-fexceptions` means we are using exceptions but it does not say which exception handling (Emscripten vs. wasm). Before we only had Emscripten EH so `-fwasm-exceptions` automatically meant Emscripten EH, but now we don't know. Currently Wasm EH is experimental and does not work fully, so still the default is Emscripten EH. So if you only specify `-fexceptions`, we use Emscripten EH, and if you either specify `-fwasm-exceptions` or both `-fexceptions` and `-fwasm-exceptions`, we use wasm EH. The plan is to switch to wasm EH as default when only `-fexceptions` is specified later, when wasm EH is fully supported in both toolchain and major VMs.

Reply:
I see, so the issue is to allow us to change which exceptions are enabled by `-fexceptions`?

I think we can still do that in a simpler way than adding these two new locals. Can't we currently have `-fexceptions` set `DISABLE_EXCEPTION_CATCHING=0` (as it already does) and in the future make it instead do `EXCEPTION_HANDLING=1`?

Reply:
We can still use the wasm EH with `-fexceptions`, if you specify `-fwasm-exceptions` as well. So to summarize:
- `-fexceptions`: Emscripten EH (but hope to switch to wasm EH)
- `-fwasm-exceptions`: Wasm EH (We need a separate flag bc `-fexceptions` defaults to Emscripten EH)
- `-fexceptions -fwasm-exceptions`: Wasm EH

Reply:
Just to throw some more random monkey wrenches in here: Is there any language-level difference between `-fexceptions` and `-fwasm-exceptions`? (currently maybe, but long-term no, right)? Would it make sense to use `-mexceptions` (i.e. the feature flag) instead? e.g. `-fexceptions` would turn on EH in the language/frontend, and `-mexceptions` would determine whether the EH is emscripten or wasm? Presumably sometime in the future `-mexceptions` would be the default, but it's possible that we might decide to always have `-fno-exceptions` be the default for emscripten since there will always be a size cost.

Reply:
Oh, I see, so `-fexceptions -fwasm-exceptions` and the same in reverse order should both enable wasm exceptions? That makes sense, and this looks like a good way to do it.

Is that documented somewhere? Maybe in `./site/source/docs/tools_reference/emcc.rst`?

Reply:
Actually that's maybe a more interesting product question than I realized: do we plan to keep an equivalent of `-s DISABLE_EXCEPTION_CATCHING` as available or as default?

Reply:
@dschuff 
> Just to throw some more random monkey wrenches in here: Is there any language-level difference between -fexceptions and -fwasm-exceptions? (currently maybe, but long-term no, right)? 

Yes. We also have `-fwasm-exceptions` in Clang, but later we can make automatically implied when a user sets `-fexceptions` and the architecture is wasm. But I don't think this will be near future.

> Would it make sense to use -mexceptions (i.e. the feature flag) instead? e.g. -fexceptions would turn on EH in the language/frontend, and -mexceptions would determine whether the EH is emscripten or wasm? Presumably sometime in the future -mexceptions would be the default, but it's possible that we might decide to always have -fno-exceptions be the default for emscripten since there will always be a size cost.

You mean, in Clang, right? I think it's better to keep `-fwasm-exceptions` there. This doesn't mean a user need to specify this `-fwasm-exceptions` every time later. If they use Emscripten, Emscripten will set it for them. Even when they want to use only Clang, we can make it implied when `-fexceptions` is specified and the archictecture is wasm. Also which one of `-fexceptions` and `-fno-exceptions` we should make the default can be changed later.

The reason I prefer to keep `-fwasm-exceptions` and make it the controlling flag over `-mexception-handling` is [here](https://reviews.llvm.org/D67208#1660213).

> Actually that's maybe a more interesting product question than I realized: do we plan to keep an equivalent of -s DISABLE_EXCEPTION_CATCHING as available or as default?

Isn't `-fignore-exceptions` for that?

@kripken 
./site/source/docs/tools_reference/emcc.rst does not seem to contain any instruction on exceptions. The current Emscripten EH is not documented there either. Also for now, wasm EH is not stable, so I'm not sure we want to add `-fwasm-exceptions` flag to the documentation. Maybe it'd be better to add `-fexceptions` as a way to turn on Emscripten EH...? I'm not sure if that is a recommended way over `-s DISABLE_EXCEPTION_CATCHING=1` for now though?

Reply:
Ah, right i had forgotten we had that discussion already. And yes, I also forgot about `-fignore-exceptions`.
And yeah I'm fine with leaving this out of the documentation until it's more fully baked. Let's just not forget to do that later.